### ISSUE TITLE
Looping commands causes "client request loop" error

### ISSUE THREAD

**Issue by diver7** [pnp/powershell#81]
### Reporting an Issue or Missing Feature
I am running a script that connects to a hub site, extracts the footer template and then applies that footer template to all the associated spoke sites for that hub site. The script runs fine without any errors and applies the footer template to the first "N" sites but then starts returning a "client request loop" error for the remainder of the sites. The error is being thrown on the Connect-PnPOnline (Line 68) command. My assumption is that the server is seeing the same commands executed over and over and assumes it is an infinite loops so it terminates the command. Not sure if you have any control over this but I wanted to report it and see if there is a fix or workaround.

### Expected behavior 
Script runs through without any errors no matter how many times I loop through the commands.

### Actual behavior 
Script starts throwing an error after some number of sites. Number of sites seems to be different each time. 
```
MSAL.NetCore.4.21.0.0.MsalUiRequiredException: 
	ErrorCode: invalid_grant
Microsoft.Identity.Client.MsalUiRequiredException: AADSTS50196: The server terminated an operation because it encountered a client request loop. Please contact your app vendor.
Trace ID: 03263d4e-a8b5-42e7-bd28-6478d8904d00
Correlation ID: 2b1c1f55-dd4e-42da-96dd-8f1fc8b3f83e
Timestamp: 2021-01-04 20:31:36Z
   at PnP.PowerShell.Commands.Base.ConnectOnline.ProcessRecord() in D:\a\powershell\powershell\src\Commands\Base\ConnectOnline.cs:line 293
   at System.Management.Automation.CommandProcessor.ProcessRecord()
	StatusCode: 400 
	ResponseBody: {"error":"invalid_grant","error_description":"AADSTS50196: The server terminated an operation because it encountered a client request loop. Please contact your app vendor.\r\nTrace ID: 03263d4e-a8b5-42e7-bd28-6478d8904d00\r\nCorrelation ID: 2b1c1f55-dd4e-42da-96dd-8f1fc8b3f83e\r\nTimestamp: 2021-01-04 20:31:36Z","error_codes":[50196],"timestamp":"2021-01-04 20:31:36Z","trace_id":"03263d4e-a8b5-42e7-bd28-6478d8904d00","correlation_id":"2b1c1f55-dd4e-42da-96dd-8f1fc8b3f83e","error_uri":"https://login.microsoftonline.com/error?code=50196"} 
	Headers: Cache-Control: no-store, no-cache
Pragma: no-cache
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
P3P: CP="DSP CUR OTPi IND OTRi ONL FIN"
client-request-id: 2b1c1f55-dd4e-42da-96dd-8f1fc8b3f83e
x-ms-request-id: 03263d4e-a8b5-42e7-bd28-6478d8904d00
x-ms-ests-server: 2.1.11328.14 - EUS ProdSlices
x-ms-clitelem: 1,50196,0,,
Set-Cookie: fpc=AnmSYBLQZ0pAgWcAHpjKSLg-FWqNFAAAADJxhdcOAAAA; expires=Wed, 03-Feb-2021 20:31:36 GMT; path=/; secure; HttpOnly; SameSite=None, x-ms-gateway-slice=estsfd; path=/; secure; httponly, stsservicecookie=estsfd; path=/; secure; samesite=none; httponly
Date: Mon, 04 Jan 2021 20:31:35 GMT
```

### Steps to reproduce behavior
```
$hubsite = Read-Host "Please enter the Hub site Url"
$global:creds = Get-Credential
$footerLogoUrl = $hubsite+"/SiteAssets/logo.png"

# Declaring and creating the log files. Each time the script is executed, a new file will be created with the current time in the filename.
$dateTime=Get-Date -format "dd-MMM-yyyy HH-mm-ss"
$ExecutionLogFile="UpdateFooter_Report"+"_"+ $dateTime + ".csv"
$ExecutionExceptionFile="UpdateFooter_Exceptions"+"_"+ $dateTime + ".log"
$Header="SiteName,Url"
Add-Content -Path $ExecutionLogFile -Value $Header

#Process hub site
$spokeSites = "spokesites.csv"
if (Test-Path $spokeSites -PathType leaf) {
	Clear-Content $spokeSites
}
$Header="SiteName,Url"
Add-Content -Path $spokeSites -Value $Header
Write-Host "Processing Hub site..."
try {
	Connect-PnPOnline -Url $hubsite -Credentials ($global:creds)
	$results = Submit-PnPSearchQuery -Query 'contentclass=sts_site' -RefinementFilters 'departmentid:string("{*",linguistics=off)' -TrimDuplicates $false -SelectProperties @("Title","Path","DepartmentId","SiteId") -All -RelevantResults
	$hubSites = $results |? { $_.DepartmentId.Trim('{','}') -eq $_.SiteId  }
	foreach( $hub in $hubSites ) {
		if ($hub.Path.TrimEnd("/") -eq $hubsite) {
			Write-Host "Getting footer from "$hub.Title"..." -NoNewLine
			Connect-PnPOnline -Url $hubsite -Credentials ($global:creds)
			Get-PnPSiteTemplate -Out .\hubfootertemplate.xml -Force -Handlers SiteFooter
			$footerparams = Get-PnPFooter
			$footerLogoPath = $hubsite+$footerparams.LogoUrl
			$footerLayout = $footerparams.Layout
			$footerBackgroundTheme = $footerparams.BackgroundTheme
			$footerTitle = $footerparams.Title
			((Get-Content -path .\hubfootertemplate.xml -Raw) -replace 'RemoveExistingNodes="false"','RemoveExistingNodes="true"') | Set-Content -Path .\hubfootertemplate.xml
			$newLogoPath = 'Logo="'+$footerLogoPath+'"'
			((Get-Content -path .\hubfootertemplate.xml -Raw) -replace 'Logo="/"',$newLogoPath) | Set-Content -Path .\hubfootertemplate.xml
			Write-Host "done." -ForegroundColor Green

			Write-Host "Getting Spoke sites for: "$hub.Title - $hubsite"..." -NoNewLine
			# Filter to sites associated to the current hub
			$associatedSites = ($results |? { $_.DepartmentId -eq $hub.DepartmentId -and $_.SiteId -ne $hub.SiteId })
			foreach($site in $associatedSites) {
				$title = (($site.Title) -replace ",","_")
				$spokesite = $title+","+ $site.Path
				Add-Content -Path $spokeSites -Value $spokesite
			}
			Write-Host "done." -ForegroundColor Green
		}
		else {
			Write-Host "Skipping Hub site: " $hub.Title " - " $hub.Path
		}
	}
}
catch [Exception] {
	Write-Host "Error." -ForegroundColor Red
	Add-Content -Path $ExecutionExceptionFile -Value $Url
	Add-Content -Path $ExecutionExceptionFile -Value "Exception processing hub site."
	Add-Content -Path $ExecutionExceptionFile -Value $_.exception
	Add-Content -Path $ExecutionExceptionFile -Value "======================================="
	Write-Host "Exiting...process incomplete!" -ForegroundColor Red
	exit	
}
#Apply footer template to spoke sites
$sitelist = Import-Csv $spokeSites
foreach ($site in $sitelist) {
	Write-Host "Updating $($site.Sitename)." -NoNewLine
	try {
		Connect-PnPOnline -Url $site.Url -Credentials ($global:creds)
		Write-Host "." -NoNewLine
		#Setting these parameters separately because was having issues when all in one command line
		Set-PnPFooter -Enabled:$true -Title $footerTitle 
		Write-Host "." -NoNewLine
		Set-PnPFooter -Layout $footerLayout
		Write-Host "." -NoNewLine
		Set-PnPFooter -BackgroundTheme $footerBackgroundTheme
		Write-Host "." -NoNewLine
		Set-PnPFooter -LogoUrl $footerLogoPath
		Write-Host "." -NoNewLine
		Invoke-PnPSiteTemplate -Path hubfootertemplate.xml
		Write-Host "." -NoNewLine
        Set-PnPWeb -SiteLogoUrl $footerLogoUrl
        Write-Host "." -NoNewLine
		$value = $site.SiteName+","+$site.Url
		Add-Content -Path $ExecutionLogFile -Value $value
		Write-Host "done." -ForegroundColor Green
	}
        catch [Exception]
        {
		Write-Host "Error." -ForegroundColor Red
                Add-Content -Path $ExecutionExceptionFile -Value $site.Url
                Add-Content -Path $ExecutionExceptionFile -Value "Exception updating the site footer"
                Add-Content -Path $ExecutionExceptionFile -Value $_.exception
                Add-Content -Path $ExecutionExceptionFile -Value "======================================="
        }
}
```

### What is the version of the Cmdlet module you are running?
0.3.25
PS 7.0.3

### Which operating system/environment are you running PnP PowerShell on?
- [X] Windows
- [ ] Linux
- [ ] MacOS
- [ ] Azure Cloud Shell
- [ ] Azure Functions
- [ ] Other : please specify

**Comment by erwinvanhunen** [comment]
Thank you for bringing this issue up! 

We do not make the requests ourselves when running the code. Behind the scenes we make use of the MSAL library to acquire access tokens, and that one is causing the looping request. We request a token and then it's added to an in-memory cache by MSAL. The next time a request for a token is made for the same resource (e.g. your tenant or the Graph) MSAL will serve the token from the cache, unless the token is expired, then it will request a new valid token and stores that in the cache for subsequent requests. Having said that: every new `Connect-PnPOnline` will reset this cache.

I just merged a change which will try to reuse this cache. The cache will live through the session of PowerShell and as long as the type of connection is the same (e.g. same credentials, or same clientid/clientsecret/clientcertificate), regardless of using Connect-PnPOnline.

Without this change I was able to reproduce your issue by doing:

```powershell
for($i=0;$i -lt 100;$i++)
{ 
    Connect-PnPOnline -Url [url] -Credentials $creds
    Get-PnPList -Identity TestList
}
```

After the change the error did not occur anymore and a fiddler trace did not show any repeated calls to the token service.

The change will be available in the nightly release that will be available tomorrow: 0.3.27-nightly
